From 9f2559ed2225e26cfe160ddab324714cec3e6167 Mon Sep 17 00:00:00 2001
From: Super User <root@localhost.localdomain>
Date: Sat, 16 May 2026 10:17:16 +0800
Subject: [PATCH] KVM: arm64: Add debug logging for HDBSS MMIO tracking

Add KVM_HDBSS_DBG prefix debug messages:
- mmu.c: MMIO fault and DBM decision
- handle_exit.c: Buffer flush entries
- arm.c: Enable/disable timing

Signed-off-by: OpenEuler Kernel Team <kernel@openeuler.org>
---
 arch/arm64/kvm/arm.c         | 9 +++++++++
 arch/arm64/kvm/handle_exit.c | 6 ++++++
 arch/arm64/kvm/mmu.c         | 9 ++++++++-
 3 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 0e4cad7532df..20748134fd7c 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -149,10 +149,13 @@ static void kvm_clear_hdbss(struct kvm *kvm)
 	struct kvm_vcpu *vcpu;
 	struct page *hdbss_pg;
 
+	pr_info("KVM_HDBSS_DBG: Disable vtcr=0x%lx\n", kvm->arch.vtcr);
+
 	kvm->arch.vtcr &= ~(VTCR_EL2_HD | VTCR_EL2_HDBSS);
 
 	kvm_for_each_vcpu(i, vcpu, kvm) {
 		/* Kick vcpus to flush hdbss buffer. */
+		pr_info("KVM_HDBSS_DBG: Disable vcpu%lu\n", i);
 		kvm_vcpu_kick(vcpu);
 
 		hdbss_pg = phys_to_page(HDBSSBR_BADDR(vcpu->arch.hdbss.br_el2));
@@ -164,6 +167,7 @@ static void kvm_clear_hdbss(struct kvm *kvm)
 	}
 
 	kvm->enable_hdbss = false;
+	pr_info("KVM_HDBSS_DBG: Disable done\n");
 }
 
 static int kvm_cap_arm_enable_hdbss(struct kvm *kvm,
@@ -190,6 +194,7 @@ static int kvm_cap_arm_enable_hdbss(struct kvm *kvm,
 	}
 
 	/* Enable the HDBSS feature if size > 0, otherwise disable it. */
+	pr_info("KVM_HDBSS_DBG: Enable req size=%d\n", size);
 	if (size) {
 		kvm->enable_hdbss = true;
 		kvm->arch.vtcr |= VTCR_EL2_HD | VTCR_EL2_HDBSS;
@@ -205,6 +210,9 @@ static int kvm_cap_arm_enable_hdbss(struct kvm *kvm,
 			vcpu->arch.hdbss.br_el2 = HDBSSBR_EL2(page_to_phys(hdbss_pg), size);
 			vcpu->arch.hdbss.prod_el2 = 0;
 
+			pr_info("KVM_HDBSS_DBG: Enable vcpu%lu br=0x%llx\n",
+				i, vcpu->arch.hdbss.br_el2);
+
 			/*
 			 * We should kick vcpus out of guest mode here to
 			 * load new vtcr value to vtcr_el2 register when
@@ -214,6 +222,7 @@ static int kvm_cap_arm_enable_hdbss(struct kvm *kvm,
 		}
 
 		kvm_info("Enable HDBSS success, HDBSS buffer size: %d\n", size);
+		pr_info("KVM_HDBSS_DBG: Enable done\n");
 	} else if (kvm->enable_hdbss) {
 		kvm_clear_hdbss(kvm);
 		kvm_info("Disable HDBSS success\n");
diff --git a/arch/arm64/kvm/handle_exit.c b/arch/arm64/kvm/handle_exit.c
index 6919c3858193..1f5af027317c 100644
--- a/arch/arm64/kvm/handle_exit.c
+++ b/arch/arm64/kvm/handle_exit.c
@@ -310,6 +310,9 @@ static void kvm_flush_hdbss_buffer(struct kvm_vcpu *vcpu)
 	isb();
 	curr_idx = HDBSSPROD_IDX(read_sysreg_s(SYS_HDBSSPROD_EL2));
 
+	pr_info_ratelimited("KVM_HDBSS_DBG: Flush start vcpu=%d entries=%d\n",
+		vcpu->vcpu_idx, curr_idx);
+
 	/* Do nothing if HDBSS buffer is empty or br_el2 is NULL */
 	if (curr_idx == 0 || vcpu->arch.hdbss.br_el2 == 0)
 		return;
@@ -328,6 +331,7 @@ static void kvm_flush_hdbss_buffer(struct kvm_vcpu *vcpu)
 			continue;
 
 		gpa = gpa & HDBSS_ENTRY_IPA_MASK;
+		pr_info_ratelimited("KVM_HDBSS_DBG: Dirty GPA[%d]=0x%llx\n", idx, gpa);
 		kvm_vcpu_mark_page_dirty(vcpu, gpa >> PAGE_SHIFT);
 	}
 
@@ -335,6 +339,8 @@ static void kvm_flush_hdbss_buffer(struct kvm_vcpu *vcpu)
 	write_sysreg_s(0, SYS_HDBSSPROD_EL2);
 	dsb(sy);
 	isb();
+
+	pr_info_ratelimited("KVM_HDBSS_DBG: Flush done\n");
 }
 #endif
 
diff --git a/arch/arm64/kvm/mmu.c b/arch/arm64/kvm/mmu.c
index fbc62b27adc8..35bdf3163010 100644
--- a/arch/arm64/kvm/mmu.c
+++ b/arch/arm64/kvm/mmu.c
@@ -1607,6 +1607,8 @@ static int user_mem_abort(struct kvm_vcpu *vcpu, phys_addr_t fault_ipa,
 		return -EFAULT;
 
 	if (kvm_is_device_pfn(pfn)) {
+		pr_info_ratelimited("KVM_HDBSS_DBG: MMIO fault ipa=0x%llx pfn=0x%lx\n",
+			fault_ipa, pfn);
 		/*
 		 * If the page was identified as device early by looking at
 		 * the VMA flags, vma_pagesize is already representing the
@@ -1678,9 +1680,14 @@ static int user_mem_abort(struct kvm_vcpu *vcpu, phys_addr_t fault_ipa,
 	if (writable)
 		prot |= KVM_PGTABLE_PROT_W;
 
+	pr_info_ratelimited("KVM_HDBSS_DBG: DBM decision ipa=0x%llx dev=%d en=%d log=%d\n",
+		fault_ipa, device, kvm->enable_hdbss, logging_active);
+
 #ifdef CONFIG_ARM64_HDBSS
-	if (kvm->enable_hdbss && logging_active)
+	if (!device && kvm->enable_hdbss && logging_active) {
 		prot |= KVM_PGTABLE_PROT_DBM;
+		pr_info_ratelimited("KVM_HDBSS_DBG: DBM added ipa=0x%llx\n", fault_ipa);
+	}
 #endif
 
 	if (exec_fault)
-- 
2.43.0

